File: U1/PROYECTO/BanquitoServer_Soap_DotNet_GR01/BanquitoServer_Soap_DotNet_GR01/Utilities/CalculadoraFinanciera.cs
using System;
using System.Collections.Generic;
using BanquitoServer_Soap_DotNet_GR01.Models;

namespace BanquitoServer_Soap_DotNet_GR01.Utilities
{
    /// <summary>
    /// Calculadora para operaciones financieras
    /// </summary>
    public static class CalculadoraFinanciera
    {
        /// <summary>
        /// Calcular la cuota fija mensual
        /// Fórmula: Cuota = Valor / (1 - ((1 + TasaPeriodo)^-NúmeroCuotas)) / TasaPeriodo
        /// </summary>
        public static decimal CalcularCuotaFija(decimal montoCredito, decimal tasaAnual, int numeroCuotas)
        {
            // Tasa mensual
            decimal tasaMensual = tasaAnual / 12;

            // (1 + tasaMensual)
            decimal unMasTasa = 1 + tasaMensual;

            // (1 + tasaMensual)^-numeroCuotas
            double potencia = Math.Pow((double)unMasTasa, -numeroCuotas);

            // 1 - potencia
            decimal numerador = 1 - (decimal)potencia;

            // numerador / tasaMensual
            decimal denominador = numerador / tasaMensual;

            // Cuota = montoCredito / denominador
            decimal cuota = montoCredito / denominador;

            return Math.Round(cuota, 2);
        }

        /// <summary>
        /// Generar la tabla de amortización completa
        /// </summary>
        public static List<CuotaAmortizacion> GenerarTablaAmortizacion(
            decimal montoCredito,
            decimal tasaAnual,
            int numeroCuotas,
            decimal cuotaFija)
        {
            List<CuotaAmortizacion> tabla = new List<CuotaAmortizacion>();

            decimal tasaMensual = tasaAnual / 12;
            decimal saldo = montoCredito;

            for (int i = 1; i <= numeroCuotas; i++)
            {
                // Calcular interés de la cuota
                decimal interes = Math.Round(saldo * tasaMensual, 2);

                // Calcular capital pagado
                decimal capitalPagado = Math.Round(cuotaFija - interes, 2);

                // Calcular nuevo saldo
                decimal nuevoSaldo = Math.Round(saldo - capitalPagado, 2);

                // Ajuste para la última cuota (por posibles diferencias de redondeo)
                if (i == numeroCuotas)
                {
                    capitalPagado = saldo;
                    cuotaFija = capitalPagado + interes;
                    nuevoSaldo = 0;
                }

                // Crear cuota de amortización
                CuotaAmortizacion cuota = new CuotaAmortizacion
                {
                    NumeroCuota = i,
                    ValorCuota = cuotaFija,
                    Interes = interes,
                    CapitalPagado = capitalPagado,
                    Saldo = nuevoSaldo
                };

                tabla.Add(cuota);

                // Actualizar saldo para la siguiente iteración
                saldo = nuevoSaldo;
            }

            return tabla;
        }

        /// <summary>
        /// Calcular la edad de una persona
        /// </summary>
        public static int CalcularEdad(DateTime fechaNacimiento)
        {
            DateTime hoy = DateTime.Today;
            int edad = hoy.Year - fechaNacimiento.Year;

            if (fechaNacimiento.Date > hoy.AddYears(-edad))
                edad--;

            return edad;
        }
    }
}

File: U2/PAO202551_ELECTRO_BDDNUBE_RESTFULL_JAVA_GR01/04. SERVIDOR/BanquitoServer_Restfull_Java_GR01/src/main/java/ec/edu/monster/Utilities/CalculadoraFinanciera.java
package ec.edu.monster.Utilities;

import ec.edu.monster.Models.CuotaAmortizacion;
import java.math.BigDecimal;
import java.math.MathContext;
import java.math.RoundingMode;
import java.time.LocalDate;
import java.time.Period;
import java.util.ArrayList;
import java.util.List;

/**
 * Calculadora para operaciones financieras
 */
public final class CalculadoraFinanciera {

    private static final MathContext MC = new MathContext(10, RoundingMode.HALF_UP);

    private CalculadoraFinanciera() {
    }

    /**
     * Calcular la cuota fija mensual
     * Fórmula: Cuota = Valor / (1 - ((1 + TasaPeriodo)^-NúmeroCuotas)) /
     * TasaPeriodo
     */
    public static BigDecimal calcularCuotaFija(BigDecimal montoCredito, BigDecimal tasaAnual, int numeroCuotas) {
        // Tasa mensual
        BigDecimal tasaMensual = tasaAnual.divide(BigDecimal.valueOf(12), MC);

        // (1 + tasaMensual)
        BigDecimal unMasTasa = BigDecimal.ONE.add(tasaMensual);

        // (1 + tasaMensual)^-numeroCuotas
        double potencia = Math.pow(unMasTasa.doubleValue(), -numeroCuotas);

        // 1 - potencia
        BigDecimal numerador = BigDecimal.ONE.subtract(BigDecimal.valueOf(potencia));

        // numerador / tasaMensual
        BigDecimal denominador = numerador.divide(tasaMensual, MC);

        // Cuota = montoCredito / denominador
        BigDecimal cuota = montoCredito.divide(denominador, MC);

        return cuota.setScale(2, RoundingMode.HALF_UP);
    }

    /**
     * Generar la tabla de amortización completa
     */
    public static List<CuotaAmortizacion> generarTablaAmortizacion(
            BigDecimal montoCredito,
            BigDecimal tasaAnual,
            int numeroCuotas,
            BigDecimal cuotaFija) {

        List<CuotaAmortizacion> tabla = new ArrayList<>();

        BigDecimal tasaMensual = tasaAnual.divide(BigDecimal.valueOf(12), MC);
        BigDecimal saldo = montoCredito;

        for (int i = 1; i <= numeroCuotas; i++) {
            // Calcular interés de la cuota
            BigDecimal interes = saldo.multiply(tasaMensual).setScale(2, RoundingMode.HALF_UP);

            // Calcular capital pagado
            BigDecimal capitalPagado = cuotaFija.subtract(interes).setScale(2, RoundingMode.HALF_UP);

            // Calcular nuevo saldo
            BigDecimal nuevoSaldo = saldo.subtract(capitalPagado).setScale(2, RoundingMode.HALF_UP);

            BigDecimal cuotaActual = cuotaFija;

            // Ajuste para la última cuota (por posibles diferencias de redondeo)
            if (i == numeroCuotas) {
                capitalPagado = saldo;
                cuotaActual = capitalPagado.add(interes);
                nuevoSaldo = BigDecimal.ZERO;
            }

            // Crear cuota de amortización
            CuotaAmortizacion cuota = new CuotaAmortizacion();
            cuota.setNumeroCuota(i);
            cuota.setValorCuota(cuotaActual);
            cuota.setInteres(interes);
            cuota.setCapitalPagado(capitalPagado);
            cuota.setSaldo(nuevoSaldo);

            tabla.add(cuota);

            // Actualizar saldo para la siguiente iteración
            saldo = nuevoSaldo;
        }

        return tabla;
    }

    /**
     * Calcular la edad de una persona
     */
    public static int calcularEdad(LocalDate fechaNacimiento) {
        return Period.between(fechaNacimiento, LocalDate.now()).getYears();
    }
}
